docs/rfc/rfc-011-delete-gas.md

```diff
@@ -0,0 +1,140 @@
+parameter.
+
+Gas has gained adoption within the Cosmos ecosystem [as part of the Cosmos SDK][cosmos-sdk-gas].
+The SDK provides facilities for tracking how much 'Gas' a transaction is expected to take
+and a mechanism for tracking how much gas a transaction has already taken.
+
+Non-SDK applications also make use of the concept of Gas. Anoma appears to implement
+[a gas system][anoma-gas] to meter the transactions it executes.
+
+While the notion of gas is present in projects that make use of Tendermint, it is
+not a concern of Tendermint's. Tendermint's value and goal is producing blocks
+via a distributed consensus algorithm. Tendermint relies on the application specific
+code to decide how to handle the transactions Tendermint has produced (or if the
+application wants to consider them at all). Gas is an application concern.
+
+Our implementation of Gas is causing issues for node operators and relayers. At
+the moment, transactions that overflow the configured 'MaxGas' can be silently rejected
+from the mempool. Overflowing MaxGas is the _only_ way that a transaction can be considered
+invalid that is not directly a result of failing the `CheckTx`. Operators, and the application,
+do not know that a transaction was removed from the mempool for this reason. A stateless check
+of this nature is exactly what `CheckTx` exists for and there is no reason for the mempool
+to keep track of this data separately.
+
+Removing Gas from Tendermint and the mempool would allow for the 'mempool' to be a better
+abstraction: any transaction that arrived at `CheckTx` and passed the check will either be
+a candidate for a later block or evicted after a TTL is reached. All other transactions are 
+completely invalid and can be discarded forever.
+
+Removing gas will not be completely straightforward. It will mean ensuring that 
+equivalent functionality can be implemented outside of the mempool using the mempool's API.
+
+## Discussion
+
+This section catalogs the functionality that will need to exist within the Tendermint
+mempool to allow Gas to be removed and replaced by application-side bookkeeping.
+
+### Requirement: Provide Mempool Tx Sorting Mechanism
+
+Gas produces a market for inclusion in a block. On many networks, a [gas fee][cosmos-sdk-fees] is
+included in pending transactions. This fee indicates how much a user is willing to
+pay per unit of execution and the fees are distributed to validators.
+
+Validators wishing to extract higher gas fees are incentivized to include transactions 
+with the highest listed gas fees into each block. This produces a naturally ordering
+of the pending transactions. Applications wishing to implement a gas mechanism need
+to be able to order the transactions in the mempool. This can trivially be accomplished 
+by sorting transactions using the `priority` field available to applications as part of
+v0.35's `ResponseCheckTx` message.
```

Comment:
I get a bit scared by this priority setting from the app and mempool usage. At this point it seems like SDK  sets the priority based on the total fees provided. Then mempool, at this point decides to evict transactions with lower priority when mempool is full and new Tx-es arrived. I mentioned this before a few times, for IBC relaying where multiple Tx-es paid from the same account (with fees according to Tx simulation) are broadcasted (sync) we need a mechanism to keep the FIFO order for the same sender. Sorting on priority alone would break the relayer.

Reply:
This problem sounds interesting but it sounds like something that the application i.e. cosmos sdk should look to handle. If messages need to be atomic or need to be executed in a certain order they should be grouped together to form a single transaction. Tendermint doesn't need to know about it. The other option is that `PrepareProposal` is used to ensure correct ordering or grouping regardless of priority.

Reply:
In v0.35 priority dictates transaction ordering in a block but in v036 priority only dictates the order that transactions are sent to the application in `PrepareProposal` not how they get to a block.

Reply:
> This problem sounds interesting but it sounds like something that the application i.e. cosmos sdk should look to handle.

Definitely. I am looking at the overall Tendermint/ SDK combo, all the ways we enhance the APIs and how we can break clients in general, relayers in particular.

> If messages need to be atomic or need to be executed in a certain order they should be grouped together to form a single transaction.

Not sure how we can do this. Most relayers are building Tx-es from events from different chains that can be issued at different times. If n events are issued for some new block on chain A, the relayer will build a single Tx with n messages and send it let's say to chain B. But if block time of A is smaller (we have seen this already), there might be new event batches from A that trigger a subsequent multi msg Tx to B. Or there could be events from yet another chain C that trigger a Tx to chain B.
In both cases we have to maintain the order because of Tx account nonce. (keeping multiple wallets on chain B, e.g. x for A->B, y for C->B, etc is an operational nightmare and still doesn't completely solve the problem) 

> In v0.35 priority dictates transaction ordering in a block ...

right, so this breaks the relayer somewhat; i think even retries triggered by the deliverTx failures won't solve much. There is a periodic cleaning mechanism that may fix this if not coming in too late.

> in v036 priority only dictates the order that transactions are sent to the application in PrepareProposal not how they get to a block.

Then if the app/SDK would want to help with this case it will need to make sure it gets all of them in case some lower nonce Tx-es ended up with lower priority.